vignettes/design_principle.Rmd

```diff
@@ -0,0 +1,196 @@
+    ii) APIs: “REDCap”, “DHIS2”, “ODK”, “Fingertips”, “goData”, “SORMAS”
+
+
+* `get_metadata()`: will be used to retrieve the data schema (table list + data dictionary) from a database or the data dictionary from an endpoint of the target API. It offers interoperability with the `clean_based_on_dictionay()` function of the [{cleanepi}](https://epiverse-trace.github.io/cleanepi/) R package.
+
+```{r get_metadata, fig.align='center', echo=FALSE}
+DiagrammeR::grViz("digraph{
+ 
+      graph[rankdir = LR]
+  
+      node[shape = rectangle, style = filled, fontname = Courier,
+      align = center]
+  
+      subgraph cluster_0 {
+        graph[shape = rectangle]
+        style = rounded
+        bgcolor = LightBlue
+    
+        label = 'get_metadata()'
+        node[shape = rectangle, fillcolor = LemonChiffon, margin = 0.25]
+        A[label = 'conn: the connection object made from authenticate()']
+        B[label = 'endpoint: the path to the target endpoint (APIs only)']
+        C[label = 'query: a list with the query parameters']
+      }
+}")
+```
+
+* `read()`: this is the main function of the package. It can be used to read from both HIS and RDBMS.
+
+```{r read, fig.align='center', echo=FALSE}
+DiagrammeR::grViz("digraph{
+ 
+      graph[rankdir = LR]
+  
+      node[shape = rectangle, style = filled, fontname = Courier,
+      align = center]
+  
+      subgraph cluster_0 {
+        graph[shape = rectangle]
+        style = rounded
+        bgcolor = LightBlue
+    
+        label = 'read()'
+        node[shape = rectangle, fillcolor = LemonChiffon, margin = 0.25]
+        A[label = 'conn: the connection object made from authenticate()']
+        B[label = 'endpoint: the path to the target endpoint (APIs only)']
+        C[label = 'query: a list with the query parameters']
+      }
+}")
+```
+
+An internal function (`read_*` where `*` is the source name) is designed for fetching data from a specific HIS. This implementation will make it possible to perform updates on a speicific function, while conserving the remaining structures (easy maintenance).
+
+```{r fig.align='center', echo=FALSE}
+DiagrammeR::DiagrammeR("
+graph TB
+    A[read]
+    A --> B(read_rdbms)
+    A --> D(read_redcap)
+    A --> E(read_dhis2)
+    A --> F(read_fingertips)
+    A --> G(read_sormas)
+")
+```
+
+Note that, when reading from RDBMS, the `query` argument could be an SQL query or a list with a vector of table names, fields and rows to subset on.
+For HIS, the elements of this list can vary depending on the API that is been queried. We strongly recommend reading the vignette on the [query_parameters](./query_parameters.Rmd) for more details.
+
+
+## Dependencies
+
+* The `read()` will rely mainly on 2 packages:
+
+  [{httr2}](https://CRAN.R-project.org/package=httr2) or [{data.tab}](https://CRAN.R-project.org/package=data.table): this will serve in constructing and performing the API requests,
```

Comment:
```suggestion
  [{httr2}](https://CRAN.R-project.org/package=httr2) or [{data.table}](https://CRAN.R-project.org/package=data.table): this will serve in constructing and performing the API requests,
```

Reply:
I accounted for this.